src/queue.rs

```diff
@@ -0,0 +1,98 @@
+use core::{
+    marker::PhantomData,
+    mem::{size_of, MaybeUninit},
+};
+
+use esp_idf_sys::TickType_t;
+
+use crate::sys;
+
+pub struct Queue<'a, T> {
+    ptr: sys::QueueHandle_t,
+    is_owned: bool,
+    _marker: PhantomData<&'a mut T>,
+}
+
+unsafe impl<'a, T> Send for Queue<'a, T> where T: Send + Sync {}
+unsafe impl<'a, T> Sync for Queue<'a, T> where T: Send + Sync {}
+
+impl<'a, T> Queue<'a, T>
+where
+    // ensures the contained elements are not `Drop`
+    // might be able to lift restriction in the future
+    T: Copy,
+{
+    pub fn new(size: usize) -> Self {
+        Queue {
+            ptr: unsafe { sys::xQueueGenericCreate(size as u32, size_of::<T>() as u32, 0) },
+            is_owned: true,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Create a new queue which is not deleted on `Drop`, but owned by somebody else.
+    ///
+    /// # Safety
+    ///
+    /// Care must be taken that the queue is valid for the constructed
+    /// lifetime.
+    pub unsafe fn new_borrowed(ptr: sys::QueueHandle_t) -> Self {
+        assert!(!ptr.is_null());
+
+        Queue {
+            ptr,
+            is_owned: false,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn as_raw(&self) -> sys::QueueHandle_t {
+        self.ptr
+    }
+
+    /// Copy item to back of queue, blocking for `timeout` ticks if full.
+    pub fn send_back(&self, item: T, timeout: TickType_t) -> bool {
```

Comment:
It would be suer-nice if we can additionally expose the `*FromISR` variants of the xQueue methods, as one of the benefits of the FreeRTOS-native synchronization primitives (including the queue) is that they can be operated on from an ISR (that is, as compared to Rust's `std::sync` primitives which are incompatible).

I mean:
- `xQueueSendFromISR`
- `xQueuePeekFromISR` 
- `xQueueReceiveFromISR`
- `xQueueSendToFrontFromISR`
- `xQueueSendToBackFromISR`
- `xQueueGiveFromISR`

You can simply retrofit your existing methods here to call the *ISR methods if `crate::interrupt::active()` returns `true`, and call what they currently call otherwise. However, for the `ISR` methods, the `timeout` parameter should be disregarded, as ISRs obviously cannot wait.

Or you can implement a new bunch of ISR-specific pub methods. I think I prefer the former.

Reply:
I would also prefer the automatic switch, however the FromISR methods accept an out parameter indicating whether or not a higher priority task was unblocked. So I will have to make separate functions.